README.md
# Python exception notifier for Airbrake

![Build Status](https://github.com/airbrake/pybrake/workflows/pybrake/badge.svg)

## Installation

pybrake requires Python 3.6+.

``` shell
pip install -U pybrake
```

## Configuration

You **must** set both `project_id` & `project_key`.

To find your `project_id` and `project_key` navigate to your project's
_Settings_ and copy the values from the right sidebar.

![project-idkey]

```python
import pybrake

notifier = pybrake.Notifier(project_id=123,
                            project_key='FIXME',
                            environment='production')
```

## Sending errors to Airbrake

```python
try:
    raise ValueError('hello')
except Exception as err:
    notifier.notify(err)
```

### Sending errors synchronously

By default, the `notify` function sends errors asynchronously using
`ThreadPoolExecutor` and returns a `concurrent.futures.Future`, a synchronous
API is also made available with the `notify_sync` function:

```python
notice = notifier.notify_sync(err)
if 'id' in notice:
    print(notice['id'])
else:
    print(notice['error'])
```

## Adding custom params

To set custom params you can build and send notice in separate steps:

```python
notice = notifier.build_notice(err)
notice['params']['myparam'] = 'myvalue'
notifier.send_notice(notice)
```

You can also add custom params to every error notice before it's sent to Airbrake
with the `add_filter` function.

```python
def my_filter(notice):
    notice['params']['myparam'] = 'myvalue'
    return notice

notifier.add_filter(my_filter)
```

## Ignoring notices

There may be some notices/errors thrown in your application that you're not
interested in sending to Airbrake, you can ignore these using the `add_filter`
function.

```python
def my_filter(notice):
    if notice['context']['environment'] == 'development':
        # Ignore notices in development environment.
        return None
    return notice

notifier.add_filter(my_filter)
```

## Filtering keys

With `keys_blocklist` option you can specify list of keys containing sensitive information that must be filtered out, e.g.:

```python
notifier = pybrake.Notifier(
    ...
    keys_blocklist=[
        'password',           # exact match
        re.compile('secret'), # regexp match
    ],
)
```

## Logging integration

pybrake provides a logging handler that sends your logs to Airbrake.

```python
import logging
import pybrake


airbrake_handler = pybrake.LoggingHandler(notifier=notifier,
                                          level=logging.ERROR)

logger = logging.getLogger('test')
logger.addHandler(airbrake_handler)

logger.error('something bad happened')
```

## Disabling pybrake logs

The pybrake logger can be silenced by setting the logging level to
`logging.CRITICAL`.

``` python
import logging


logging.getLogger("pybrake").setLevel(logging.CRITICAL)
```

## Sending route stats

`notifier.routes.notify` allows sending route stats to Airbrake. The library
provides integrations with Django and Flask. (your routes are tracked
automatically). You can also use this API manually:

```py
from pybrake import RouteMetric

metric = RouteMetric(method=request.method, route=route)
metric.status_code = response.status_code
metric.content_type = response.headers.get("Content-Type")
metric.end_time = time.time()

notifier.routes.notify(metric)
```

## Sending route breakdowns

`notifier.routes.breakdowns.notify` allows sending performance breakdown stats
to Airbrake. You can use this API manually:

```py
from pybrake import RouteMetric

metric = RouteMetric(
    method=request.method,
    route='/things/1',
    status_code=200,
    content_type=response.headers.get('Content-Type'))
metric._groups = {'db': 12.34, 'view': 56.78}
metric.end_time=time.time()

notifier.routes.breakdowns.notify(metric)
```

## Sending query stats

`notifier.queries.notify` allows sending SQL query stats to Airbrake. The
library provides integration with Django (your queries are tracked
automatically). You can also use this API manually:

```py
notifier.queries.notify(
    query="SELECT * FROM foos",
    method=request.method,
    route=route,
    function="test",
    file="test",
    line=10,
    start_time=time.time(),
    end_time=time.time(),
)
```

## Sending queue stats

`notifier.queues.notify` allows sending queue (job) stats to Airbrake. The
library provides integration with Celery (your queues are tracked
automatically). You can also use this API manually:

```py
from pybrake import QueueMetric

metric = QueueMetric(queue="foo_queue")
metric._groups = {'redis': 24.0, 'sql': 0.4}
notifier.queues.notify(metric)
```

## Framework Integration

Pybrake provides a ready-to-use solution with minimal configuration for python 
frameworks.

* [AIOHTTP](https://docs.airbrake.io/docs/platforms/framework/python/aiohttp)
* [BottlePy](https://docs.airbrake.io/docs/platforms/framework/python/bottle)
* [Celery](https://docs.airbrake.io/docs/platforms/framework/python/celery)
* [CherryPy](https://docs.airbrake.io/docs/platforms/framework/python/cherrypy)
* [Django](https://docs.airbrake.io/docs/platforms/framework/python/django)
* [FastAPI](https://docs.airbrake.io/docs/platforms/framework/python/fastapi)
* [Falcon](https://docs.airbrake.io/docs/platforms/framework/python/falcon)
* [Flask](https://docs.airbrake.io/docs/platforms/framework/python/flask)
* [Hug](https://docs.airbrake.io/docs/platforms/framework/python/hug)
* [Masonite](https://docs.airbrake.io/docs/platforms/framework/python/masonite)
* [Pycnic](https://docs.airbrake.io/docs/platforms/framework/python/pycnic)
* [Pyramid](https://docs.airbrake.io/docs/platforms/framework/python/pyramid)
* [Sanic](https://docs.airbrake.io/docs/platforms/framework/python/sanic)
* [Starlette](https://docs.airbrake.io//docs/platforms/framework/python/starlette)
* [Tornado](https://docs.airbrake.io//docs/platforms/framework/python/tornado)
* [Turbogears2](https://docs.airbrake.io//docs/platforms/framework/python/turbogears2)

## Development

### Running the tests

```shell
pip install -r requirements.txt
pip install -r test-requirements.txt
pytest
```

### Uploading to PyPI

```shell
python setup.py sdist upload
```

### Remote configuration

Every 10 minutes the notifier issues an HTTP GET request to fetch remote
configuration. This might be undesirable while running tests. To suppress this
HTTP call, you need to pass `remote_config=False` to the notifier.

[project-idkey]: https://s3.amazonaws.com/airbrake-github-assets/pybrake/project-id-key.png


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.10.1] - 2023-01-10

### Changes

- Started handling HTTP response 429 from the Airbrake Notice API more 
  granularly. Previously, we would return IP is rate limited whenever the 
  Airbrake Notice API returned a 429. However, this response code may also indicate an error cap hit ([#254])

## [1.10.0] - 2022-10-20

### Added

- Added support for the Morepath framework with a sample example

### Changes

- Query statistics improved to pass function, file, and line number


## [1.9.0] - 2022-09-27

### Added

- Added support for the Tornado framework with a sample example
- Added support for the Turbogears2 framework with a sample example


## [1.8.0] - 2022-09-13

### Added

- Added support for the Pycnic framework with a sample example
- Added support for the Starlette framework with a sample example

### Changes

- Receives notices and APM events and periodically sends them
  asynchronously in the background (1 minute). Up to 100 backlog items for
  notice and APM Performance (Shared for Routes, Route breakdown, query,
  queue stats). In the event that the backlog queue becomes full, the item
  will be replaced and an error will be recorded.

## [1.7.0] - 2022-09-02

### Added

- Add support for Falcon framework with sample example
- Add support for Hug framework with sample example
- Add queue for backlog thread for failed request data

### Fixed

- Fastapi/Sanic middleware does not support lack of SqlAlchemy package [229](https://github.com/airbrake/pybrake/issues/229)

## [1.6.0] - 2022-07-08

### Added

- Add sample example for AIOHTTP Framework

### Changed

- Improve AIOHTTP middleware with APM support of Jinja Template, SqlAlchemy

### Fixed

- Fix for AIOHTTP middleware's "none type object has no attribute while handling error" 

## [1.5.0] - 2022-06-07

### Added

- Masonite Middleware with sample example

### Changed

- Bump bottle from 0.12.19 to 0.12.20 in /examples/bottle

## [1.4.0] - 2022-05-31

### Added

- Sanic Middleware
- Sample example for Pyramid, CherryPy, Sanic, & FastAPI middleware

## [1.3.0] - 2022-05-26

### Added

- Changelog file
- Pyramid middleware
- CherryPy middleware
- FastAPI middleware

## [1.2.0] - 2022-05-11

### Added

- Bottle middleware
- Bottle sample example

### Fixed

- Errors are filtered out when middleware is involved in backtrace [#191](https://github.com/airbrake/pybrake/issues/191)

## [1.1.0] - 2022-05-09

### Changed

- Package Restructure
- Add Django sample example
- Update Flask sample example

### Fixed

- Fix query stats send to the Airbrake

## [1.0.6] - 2022-04-13

### Changed

- Improve pytest

### Fixed

- 400 Bad Request: Did not attempt to load JSON data because the request
  Content-Type was not 'application/json' on Flask request filter [#183](https://github.com/airbrake/pybrake/issues/183)
- Fix pylint

## [1.0.5] - 2022-03-23

### Changed

- Improve test reliability by cleaning state after running
  test_celery_integration
- README: tweak breakdowns docs, queue example to include groups, indent with 4 spaces instead of 2 python style

### Fixed

- Fixed pylint errors and disabled argument for import warning
- Flask integration order of filters does not apply blocklist to request
  object [#132](https://github.com/airbrake/pybrake/issues/132)
- blocklist_filter attempts to change Immutable [#133](https://githubcom/airbrake/pybrake/issues/133)

## [1.0.4] - 2021-06-07

### Changed

- notifier: make sure performance_stats are always True

### Fixed

- remote_settings: don't overwrite hosts with None:- When `apm_host` or 
  `error_host` is undefined, it'll overwrite whatever value is in the 
  current remote config. We don't want that to happen.
- .pylintrc: disable 'consider-using-with':- This is just noise at the time 
  of writing this commit, but we should probably address it in the future.
- Pybrake query notifier throws error when query is a psycopg2 Composed
  object [#156](https://github.com/airbrake/pybrake/issues/156)
- Fixes exception for unhashable type: 'Composed' [#157](https://github.com/airbrake/pybrake/pull/157)

## [1.0.3] - 2021-04-13

### Fixed

- Flask Middleware :- Return response when performance_stats is false

## [1.0.2] - 2021-02-16

### Removed

- remote_settings: delete print that's left from debugging

## [1.0.1] - 2021-02-16

### Changed

- Change naming style for remote_settings, settings_data

### Fixed

- remote_settings: fix fetching config route (The config route setting has no 
effect due to a bug where pybrake doesn't read that value prior to making a 
GET request to the notifier config server. The fix is to call `config_route()`
on every notifier config GET call. Additionally, we are adding a logic that 
prevents remote settings to crash the background thread in case `config_route` 
is a bad value (HTTP lib would raise an error). When an error happens such as
403 Forbidden, we make a 2nd request to the old config route, which was known
to work)

## [1.0.0] - 2021-01-21

### Added

- Pybrake SDK
- Middleware for Flask, Django, celery, aiohttp frameworks

[Unreleased]: https://github.com/airbrake/pybrake/compare/v1.10.1...HEAD
[1.10.1]: https://github.com/airbrake/pybrake/compare/v1.10.0...v1.10.1
[1.10.0]: https://github.com/airbrake/pybrake/compare/v1.9.0...v1.10.0
[1.9.0]: https://github.com/airbrake/pybrake/compare/v1.8.0...v1.9.0
[1.8.0]: https://github.com/airbrake/pybrake/compare/v1.7.0...v1.8.0
[1.7.0]: https://github.com/airbrake/pybrake/compare/v1.6.0...v1.7.0
[1.6.0]: https://github.com/airbrake/pybrake/compare/v1.5.0...v1.6.0
[1.5.0]: https://github.com/airbrake/pybrake/compare/v1.4.0...v1.5.0
[1.4.0]: https://github.com/airbrake/pybrake/compare/v1.3.0...v1.4.0

pybrake/queries.py
import base64
import json
import threading
import urllib.request
import urllib.error

from . import metrics
from .tdigest import TDigestStat, as_bytes
from .utils import logger, time_trunc_minute


class QueryStat(TDigestStat):

    def __new__(cls, *, query="", method="", route="", time):
        instance = super(QueryStat, cls).__new__(cls)
        instance.__slots__ = instance.__slots__ + (
            "query", "method", "route", "time"
        )
        return instance

    @property
    def __dict__(self):
        tdigest = as_bytes(self.td)
        self.tdigest = base64.b64encode(tdigest).decode("ascii")

        return {s: getattr(self, s) for s in self.__slots__ if s != "td"}

    def __init__(self, *, query="", method="", route="", time=None):
        super().__init__()
        self.query = query
        self.method = method
        self.route = route
        self.time = time_trunc_minute(time)


class QueryStats:
    def __init__(self, *, project_id=0, project_key="", **kwargs):
        self._config = kwargs["config"]

        self._project_id = project_id
        self._ab_headers = {
            "Content-Type": "application/json",
            "Authorization": "Bearer " + project_key,
        }
        self._env = kwargs.get("environment")

        self._thread = None
        self._lock = threading.Lock()
        self._stats = None

    def notify(self, *, query="", method="", route="", start_time=None, end_time=None):
        if not self._config.get("performance_stats"):
            return
        if not self._config.get("query_stats"):
            return

        key = query_stat_key(query=query, method=method, route=route, time=start_time)
        ms = (end_time - start_time) * 1000

        with self._lock:
            if self._stats is None:
                self._stats = {}
                self._thread = threading.Timer(metrics.FLUSH_PERIOD, self._flush)
                self._thread.start()

            if key in self._stats:
                stat = self._stats[key]
            else:
                stat = QueryStat(
                    query=query, method=method, route=route, time=start_time
                )
                self._stats[key] = stat
            stat.add(ms)

    def _flush(self):
        stats = None
        with self._lock:
            stats = self._stats
            self._stats = None

        if not stats:
            raise ValueError("stats is empty")

        out = {"queries": [v.__dict__ for v in stats.values()]}
        if self._env:
            out["environment"] = self._env

        out = json.dumps(out).encode("utf8")
        req = urllib.request.Request(
            self._ab_url(), data=out, headers=self._ab_headers, method="POST"
        )

        try:
            resp = urllib.request.urlopen(req, timeout=5)
        except urllib.error.HTTPError as err:
            resp = err
        except Exception as err:  # pylint: disable=broad-except
            logger.error(err)
            return

        try:
            body = resp.read()
        except IOError as err:
            logger.error(err)
            return

        if 200 <= resp.code < 300:
            return

        if not 400 <= resp.code < 500:
            err = f"airbrake: unexpected response status_code={resp.code}"
            logger.error(err)
            return

        if resp.code == 429:
            return

        try:
            body = body.decode("utf-8")
        except UnicodeDecodeError as err:
            logger.error(err)
            return

        try:
            in_data = json.loads(body)
        except ValueError as err:  # json.JSONDecodeError requires Python 3.5+
            logger.error(err)
            return

        if "message" in in_data:
            logger.error(in_data["message"])
            return

    def _ab_url(self):
        return f"{self._config.get('apm_host')}/api/v5/projects/{self._project_id}/queries-stats"


def query_stat_key(*, query="", method="", route="", time=None):
    time = time // 60 * 60
    return (query, method, route, time)
